src/client/dfuse/src/il/intercept.h

```diff
@@ -0,0 +1,146 @@
+ * DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
+ * THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+#ifndef __INTERCEPT_H__
+#define __INTERCEPT_H__
+#include <unistd.h>
+#include <stdlib.h>
+#include <sys/uio.h>
+#include "log.h"
+#include "ios_gah.h"
+#include "iof_fs.h"
+#include "iof_io.h"
+#include "iof_api.h"
+
+/* Low level I/O functions we intercept
+ *
+ * We purposefully skip the following:
+ * fileno
+ * fileno_unlocked
+ * sync
+ * munmap
+ * msync
+ * mremap
+ * select
+ * all aio routines (for now)
+ * fcntl (for now though we likely need for dup)
+ */
+#define FOREACH_ALIASED_INTERCEPT(ACTION)                                     \
+	ACTION(FILE *,  fopen,     (const char *, const char *))              \
+	ACTION(FILE *,  freopen,   (const char *, const char *, FILE *))      \
+	ACTION(int,     open,      (const char *, int, ...))                  \
+	ACTION(ssize_t, pread,     (int, void *, size_t, off_t))              \
+	ACTION(ssize_t, pwrite,    (int, const void *, size_t, off_t))        \
+	ACTION(off_t,   lseek,     (int, off_t, int))                         \
+	ACTION(ssize_t, preadv,    (int, const struct iovec *, int, off_t))   \
+	ACTION(ssize_t, pwritev,   (int, const struct iovec *, int, off_t))   \
+	ACTION(void *,  mmap,      (void *, size_t, int, int, int, off_t))
+
+#define FOREACH_SINGLE_INTERCEPT(ACTION)                                      \
+	ACTION(int,     fclose,    (FILE *))                                  \
+	ACTION(int,     close,     (int))                                     \
+	ACTION(ssize_t, read,      (int, void *, size_t))                     \
+	ACTION(ssize_t, write,     (int, const void *, size_t))               \
+	ACTION(ssize_t, readv,     (int, const struct iovec *, int))          \
+	ACTION(ssize_t, writev,    (int, const struct iovec *, int))          \
+	ACTION(int,     fsync,     (int))                                     \
+	ACTION(int,     fdatasync, (int))                                     \
+	ACTION(int,     dup,       (int))                                     \
+	ACTION(int,     dup2,      (int, int))                                \
+	ACTION(int,     fcntl,     (int fd, int cmd, ...))                    \
+	ACTION(FILE *,  fdopen,    (int, const char *))
+
+#define FOREACH_INTERCEPT(ACTION)            \
+	FOREACH_SINGLE_INTERCEPT(ACTION)     \
+	FOREACH_ALIASED_INTERCEPT(ACTION)
+
+#ifdef IOIL_PRELOAD
+#include <dlfcn.h>
+
+#define IOIL_FORWARD_DECL(type, name, params)  \
+	static type(*__real_##name) params;
+
+#define IOIL_DECL(name) name
+
+#define IOIL_DECLARE_ALIAS(type, name, params) \
```

Comment:
(style)  Macros with multiple statements should be enclosed in a do - while loop

(style)  macros should not use a trailing semicolon